### ISSUE TITLE
Support early const eval for inline `const{}` expressions

### ISSUE THREAD

**Issue by BergmannAtmet** [rust-lang/rust-analyzer#21080]
Related:
https://rust-lang.zulipchat.com/#narrow/channel/131828-t-compiler/topic/const.20binding.20vs.2E.20const.7B.7D.20evaluation/near/558550788

Const eval is not guaranteed to be done at `check` time, especially with `const{}` expressions.

For example:

``` rust
// check error
const X: NonZeroU32 = NonZeroU32::new(0).unwrap();
// check success, build error
let x: NonZeroU32 = const{NonZeroU32::new(0).unwrap()};
```

`rust-analyzer` catching such errors would be a great addition.

-----

I hope this is not a duplicate, since it's not easy to search for this.

**Comment by Veykril** [comment]
Not really a dupe but our const eval is very bare bones so we don't catch a lot of things in the first place. We also don't evaluate constants in a bunch of contexts yet (this one included)
